File: util/sim-event-processor/src/edu/colorado/phet/simeventprocessor/EventLog.java
// Copyright 2002-2011, University of Colorado
package edu.colorado.phet.simeventprocessor;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.StringTokenizer;

import edu.colorado.phet.common.phetcommon.util.ObservableList;
import edu.colorado.phet.common.phetcommon.util.function.Function1;

/**
 * State for a single parse run.
 *
 * @author Sam Reid
 */
public class EventLog implements Iterable<Entry> {
    private String machineID;
    private String sessionID;
    private long serverTime;
    private List<Entry> lines = new ArrayList<Entry>();

    public EventLog() {
    }

    public EventLog( String machineID, String sessionID, long serverTime, List<Entry> lines ) {
        this.machineID = machineID;
        this.sessionID = sessionID;
        this.serverTime = serverTime;
        this.lines = lines;
    }

    //Parse a single line
    public void parseLine( String line ) {
        if ( line.startsWith( "machineID" ) ) {
            machineID = readValue( line );
        }
        else if ( line.startsWith( "sessionID" ) ) {
            sessionID = readValue( line );
        }
        else if ( line.startsWith( "serverTime" ) ) {
            serverTime = Long.parseLong( readValue( line ) );
        }
        else {
            Entry entry = Entry.parse( line );
            lines.add( entry );
        }
    }

    //Read the third token, the value in a "name = value" list
    private String readValue( String line ) {
        StringTokenizer stringTokenizer = new StringTokenizer( line, " " );
        stringTokenizer.nextToken();//machineID
        stringTokenizer.nextToken();//=
        return stringTokenizer.nextToken();
    }

    public Iterator<Entry> iterator() {
        return lines.iterator();
    }

    public EventLog removeSystemEvents() {
        return removeItems( new Function1<Entry, Boolean>() {
            public Boolean apply( Entry entry ) {
                return entry.object.toLowerCase().startsWith( "system" );
            }
        } );
    }

    public EventLog removeItems( Function1<Entry, Boolean> filter ) {
        return new EventLog( machineID, sessionID, serverTime, new ObservableList<Entry>( lines ).removeItems( filter ) );
    }

    public long getLastTime() {
        return lines.get( lines.size() - 1 ).time;
    }

    public int getNumberOfEvents( final long time ) {
        EventLog log = removeItems( new Function1<Entry, Boolean>() {
            public Boolean apply( Entry entry ) {
                return entry.time > time;
            }
        } );
        EventLog user = log.removeSystemEvents();
        return user.size();
    }

    private int size() {
        return lines.size();
    }
}

File: util/sim-event-processor/src/edu/colorado/phet/simeventprocessor/scala/EventLog.scala
// Copyright 2002-2011, University of Colorado
package edu.colorado.phet.simeventprocessor.scala

import java.util.Date
import java.text.SimpleDateFormat
import collection.mutable.HashMap
import scala.collection.JavaConversions._
import edu.colorado.phet.simeventprocessor.{JavaEntry, JavaEventLog}
import edu.colorado.phet.common.phetcommon.util.function.Function1
import java.lang.Boolean

/**
 * Adds scala-convenient interface for REPL.
 * @author Sam Reid
 */
class EventLog(log: JavaEventLog) {
  val machine = log.getMachineID
  val simName = log.getSimName
  val epoch = log.getServerStartTime
  val simVersion = log.getSimVersion
  val study = log.getStudy
  val size = log.size
  val session = log.getSessionID
  val user = log.getID
  val startDate = new Date(epoch)
  val day = new SimpleDateFormat("MM-dd-yyyy").format(startDate)
  val lastTime = log.getLastTime

  def countEvents(until: Long):Int = log.getNumberOfEvents(until)

  def countMatches(matcher:Seq[Match], until:Long):Int = {
    val earlyEnoughEvents = log.filter(_.timeMilliSec < until)
    matcher.filter( matchItem => earlyEnoughEvents.find( matchItem.matches( _ )).isDefined).size
  }

  override def toString = simName + " " + simVersion + " " + new Date(epoch) + " (" + epoch + "), study = " + study + ", user = " + user + ", events = " + size + ", machineID = " + machine + ", sessionID = " + session

  lazy val histogramByObject = {
    val map = new HashMap[String, Int]
    for ( elm: JavaEntry <- log ) {
      val currentValue = map.getOrElse(elm.actor, 0)
      map.put(elm.actor, currentValue + 1)
    }
    map
  }

  def matches(entry:Match):Boolean = log.filter(entry.matches(_)).size>0

  def findMatches(matcher: Seq[Match]):List[Match]= (for ( m <- matcher if matches(m) ) yield m).toList

  def findEvents(matcher:Match):List[JavaEntry] = log.filter( matcher.matches(_)).toList
  def findEvents(actor:String):List[JavaEntry] = log.filter( ActorRule(actor).matches(_)).toList
}